initramfs-tools/scripts/init-premount/tcg_opal

```diff
@@ -42,30 +49,87 @@ get_block_device_info()
 	fi
 }
 
+opal_unlock_all_ranges()
+{
+	local device
+        local pass
+	local tmpfile
+	local list_out
+	device="$1"
+	pass="$2"
+	tmpfile=/tmp/sedutil-listlockingranges
+	# Collect a list of WLocked values for each locking range
+	"$SEDUTIL" --listLockingRanges "$pass" "$device" | gawk 'match($0, /RLKEna = [YN]  WLKEna = (Y|N)  RLocked = [YN]  WLocked = (Y|N)/, cap) { print cap[1], cap[2] }' > $tmpfile
+
+	list_out="$(cat $tmpfile)"
+	if [ -z "$list_out" ]; then
+		# file is empty means sedutil failed
+		rm -f $tmpfile
+		return 1
+	fi
+	
+	local current_range
+	local failure
+	local wlocked
+	current_range=-1
+	failure=""
+	# unlock each range that has "WLocked = Y"
+	while read wlkena wlocked; do
+		current_range=$((current_range + 1))
+		if [ "$wlkena" = "N" ]; then
+			# write lock not enabled, ignore this locking range
+			continue
+		fi
+		if [ "$wlocked" = "Y" ]; then
+			# write lock enabled and locked
+			if ! "$SEDUTIL" --setlockingrange $current_range rw "$pass" "$device" > /dev/null 2> /dev/null; then
+				failure="unexpected: unlocking failed for $device"
+				break
+			fi
+		fi
+		# always invoke prepareForS3Sleep even if the device is not locked (reboot).
+		# this is necessary to store the password in kernel memory and
+		# allow wake from sleep
+		if ! "$SEDUTIL" --prepareForS3Sleep $current_range "$pass" "$device"; then
+			failure="unexpected: prepare for S3 failed for $device"
+			break
+		fi
+	done < $tmpfile
+	rm -f $tmpfile
+	if [ -n "$failure"  ] ; then
+		panic "$failure"
+	fi
+}
+
 opal_unlock()
 {
+	local device
+	device="$1"
 	while true ; do
-		/usr/bin/stty -echo
-		echo -n "Password for $2 ($1): "
-		read pass
-		/usr/bin/stty echo
-		"$SEDUTIL" --setlockingrange 1 rw "$pass" "$1"
-		if [ $? != 0 ] ; then
+		if [ -z "$pass" ]; then
+			# Only ask if "pass" is not set. This allows the
+			# user to define the same password for multiple drives
+			# and only be prompted once.
+			echo
+			/usr/bin/stty -echo
+			echo -n "Password for $2 ($device): "
+			read pass
+			/usr/bin/stty echo
+			echo
+		fi
+		if ! opal_unlock_all_ranges "$device" "$pass"; then
+			pass=""
 			log_failure_msg "Failed to unlock!"
```

Comment:
I believe this will yield an error every time, if the user has 2 drivers with different passwords.